File: 02-two_pointers/00125-Valid_Palindrome/00125-valid_palindrome.go
// 125: Valid Palindrome
// https://leetcode.com/problems/valid-palindrome/

package main

import (
	"fmt"
	"regexp"
	"strings"
)

func isalnum(c string) bool {
    return regexp.MustCompile(`^[a-zA-Z0-9]+$`).MatchString(c)
}

// SOLUTION
func isPalindrome(s string) bool {
	l := 0;
	r := len(s) - 1

	for l<r {
		for (!isalnum(string(s[l])) && l<r) {l++}
		for (!isalnum(string(s[r])) && l<r) {r--}
		if strings.ToLower(string(s[l])) != strings.ToLower(string(s[r])) {return false}

		l++
		r--
	}

	return true
}


func main() {
	// INPUT
	s := "race a car"

	// OUTPUT
	result := isPalindrome(s)
	fmt.Println(result)
}

File: 02-two_pointers/00125-Valid_Palindrome/00125-valid_palindrome.cpp
// 125: Valid Palindrome
// https://leetcode.com/problems/valid-palindrome/


#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    // SOLUTION
    bool isPalindrome(string s) {
        int l = 0;
        int r = s.size() - 1;

        while (l<r) {
            while (!isalnum(s[l]) && l<r) l++;
            while (!isalnum(s[r]) && l<r) r--;
            if (tolower(s[l]) != tolower(s[r])) return false;
            
            l++;
            r--;
        }

        return true;
    }
};


int main() {
    Solution o;
    
    // INPUT
    string s = "race a car";

    // OUTPUT
    auto result = o.isPalindrome(s);
    cout<<(result ? "true" : "false")<<endl;

    return 0;
}